src/witness_generator/machines/fixed_lookup_machine.rs

```diff
@@ -14,9 +14,72 @@ use crate::witness_generator::{
     EvalResult, FixedData,
 };
 
+/// Indices for sets of fixed columns. For each set `COLUMNS`, stores `(V, rows)` iff every `row` in `rows` satisfies `COLUMNS[row] == V`
+#[derive(Default)]
+pub struct IndexedColumns {
+    indices: HashMap<Vec<String>, BTreeMap<Vec<AbstractNumberType>, Vec<DegreeType>>>,
+}
+
+impl IndexedColumns {
+    /// get the rows at which the assignment is satisfied
+    /// Warning: in particular, an empty assignment will return all rows
+    fn get_matches(
+        &mut self,
+        fixed_data: &FixedData,
+        mut assignment: Vec<(String, AbstractNumberType)>,
+    ) -> Option<&Vec<DegreeType>> {
+        // sort in order to have a single index for [X, Y] and for [Y, X]
+        assignment.sort_by(|(name0, _), (name1, _)| name0.cmp(name1));
+        let (input_fixed_columns, values): (Vec<_>, Vec<_>) = assignment.into_iter().unzip();
+
+        self.ensure_index(fixed_data, &input_fixed_columns);
+
+        // get the rows at which the input matches
+        self.indices
+            .get(&input_fixed_columns)
+            .as_ref()
+            .unwrap()
+            .get(&values)
+    }
+
+    /// Create an index for a set of rows, if does not exist already
+    /// `input_fixed_columns` is assumed to be sorted
+    fn ensure_index(&mut self, fixed_data: &FixedData, sorted_input_fixed_columns: &Vec<String>) {
+        // we do not use the Entry API here because we want to clone `input_fixed_columns` only on index creation
+        if self.indices.get(sorted_input_fixed_columns).is_none() {
+            // create index for this lookup
+
+            // get all values for the columns to be indexed
+            let key_values = sorted_input_fixed_columns
```

Comment:
maybe
```suggestion
            let col_ids = sorted_input_fixed_columns
```

Reply:
I want to change this to `column_values` as we're getting the concrete values here

Reply:
But these are column ids and not actual values as in `col_values[row]`, right?